staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go

```diff
@@ -21,56 +21,128 @@ import (
-	// [KubeClient] or [NodeName] options.
+	// At the time when this gets called, the original ResourceClaims may have
+	// been deleted already. They also don't get cached by the kubelet. Therefore
+	// parameters for each ResourcClaim are only the UID, namespace and name.
+	// It is the responsibility of the DRA driver to cache whatever additional
+	// information it might need about prepared resources.
 	//
-	// The caller may modify the resources after this call returns.
-	PublishResources(ctx context.Context, resources resourceslice.DriverResources) error
+	// This call must be idempotent because the kubelet might have to ask
+	// for un-preparation multiple times, for example if it gets restarted.
+	// Therefore it is not an error if this gets called for a ResourceClaim
+	// which is not currently prepared.
+	//
+	// As with PrepareResourceClaims, the helper takes care of logging
+	// and serializaton.
+	//
+	// The conventions for returning one overall error and several per-ResourceClaim
+	// errors are the same as in PrepareResourceClaims.
+	UnprepareResourceClaims(ctx context.Context, claims []NamespacedObject) (result map[types.UID]string, err error)
+}
 
-	// This unexported method ensures that we can modify the interface
-	// without causing an API break of the package
-	// (https://pkg.go.dev/golang.org/x/exp/apidiff#section-readme).
-	internal()
+// PrepareResult contains the result of preparing one particular ResourceClaim.
+type PrepareResult struct {
+	// Error, if non-empty, describes a problem that occurred while preparing
+	// the ResourceClaim. The devices are then ignored and the kubelet will
+	// try to prepare the ResourceClaim again later.
+	Error string
+
+	// Devices contains the IDs of CDI device associated with specific requests
+	// in a ResourceClaim. Those IDs will be passed on to the container runtime
+	// by the kubelet.
+	//
+	// The empty slice is also valid.
+	Devices []Device
+}
+
+// Device provides the CDI device IDs for one request in a ResourceClaim.
+type Device struct {
+	// Requests lists the names of requests or subrequests in the
+	// ResourceClaim that this device is associated with. The subrequest
+	// name may be included here, but it is also okay to just return
+	// the request name.
+	//
+	// A DRA driver can get this string from the Request field in
+	// [resourceapi.DeviceRequestAllocationResult], which includes the
+	// subrequest name if there is one.
+	//
+	// If empty, the device is associated with all requests.
+	Requests []string
+
+	// PoolName identifies the DRA driver's pool which contains the device.
+	// Must not be empty.
+	PoolName string
+
+	// DeviceName identifies the device inside that pool.
+	// Must not be empty.
+	DeviceName string
+
+	// CDIDeviceIDs lists all CDI devices associated with this DRA device.
+	// Each ID must be of the form "<vendor ID>/<class>=<unique name>".
+	// May be empty.
+	CDIDeviceIDs []string
```

Comment:
Question: Does it make sense to wrapp the id's in a type for parsing etc?

Reply:
Nothing in Kubernetes parses these IDs. If https://github.com/cncf-tags/container-device-interface had a type for it *and* we could import that package, we could use that type. But that is a lot of ifs for a little bit of extra type safety, so perhaps let's stick with `string`?